### PR TITLE
move version map register to begin block

### PR DESCRIPTION
**Pull request by nghuyenthevinh2000**
## Summary of changes

Move version map to BeginBlocker

## Report of required housekeeping

- [ ] Github issue OR spec proposal link
- [ ] Wrote tests
- [ ] Updated API documentation (client/lcd/swagger-ui/swagger.yaml)
- [ ] Added a relevant changelog entry: clog add [section] [stanza] [message]

----

## (FOR ADMIN) Before merging

- [ ] Added appropriate labels to PR
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" (coding standards)
- [ ] Confirm added tests are consistent with the intended behavior of changes
- [ ] Ensure all tests pass

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -11,7 +11,7 @@
 
 # Build
 bin
-build
+_build
```
Comment (hanjukim) [inline .gitignore:14]: It's good to not make unrelated changes in the PR. It is not mandatory though.

**Comment by inon-man** [comment]
Why does the replacement required? It's weird to me that setting version map is required at every new block. Which is different from https://github.com/cosmos/cosmos-sdk/blob/5213bbd65326edb9cd84f70c1bc0232b7389a4f7/simapp/app.go#L580

**Comment by ZaradarBH** [comment]
I agree with @inon-man. This doesnt seem right. I think we should all get together and talk  about the upgrade  handlers before we get to fare ahead of ourselves on the subject. :)

**Comment by nghuyenthevinh2000** [comment]
after Edward shows me that InitChain() will not be reached if we continue the chain and not from block zero. I kinda suspect whether continuing from 1.0.4 to 1.0.5 will reach InitChainer(). If so, the version map registration will never be reached if it is in InitChainer()

Also from Tendermint here, https://github.com/tendermint/tendermint/blob/main/consensus/replay.go#L303, RequestInitChain only happens if appBlockHeight == 0. abci.RequestInitChain{} request will invoke InitChainer().

For appBlockHeight > 0, abci.RequestInitChain{} request will not be sent.

**Comment by inon-man** [comment]
@nghuyenthevinh2000 I see. It seems that we only need to do it once in the hotfix.
I propose to keep the old SetModuleVersionMap in InitChainer to be consistent with the official cosmos-sdk method and remove BeginBlocker patch in the next upgrade.

File: app/app.go [hunk 2]
```diff
@@ -626,6 +626,12 @@ func (app *TerraApp) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock) a
 		}
 	}
 
+	// check if version map in upgrade keeper is nil
+	vm := app.UpgradeKeeper.GetModuleVersionMap(ctx)
```
Comment (ZaradarBH) [inline app/app.go:630]: Don't we need to wrap this logic in a "blockheight guard"? As such GetModuleVersionMap calls the underlying KVStore @ https://github.com/cosmos/cosmos-sdk/blob/f2b6013cee0fd4cd0d3751a173f1344cdd718e67/x/upgrade/keeper/keeper.go#L155 so I suppose this will generate another transaction that we dont want triggering before a certain target block height or I suppose it would alter the TX  signatures when nodes use the 1.0.5 client to sync from columbus-5 genesis? :)

Comment (inon-man) [reply app/app.go:630]: It is `reading` values from the store. not making a tx

Comment (inon-man) [reply app/app.go:630]: You are partially right. I think there should be a height guard for `SetModuleVersionMap` which changes the state of the chain.

Comment (edk208) [reply app/app.go:630]: ok will have to wrap in block height guards then...

**Review by edk208 — APPROVED** [review]

**Comment by nghuyenthevinh2000** [comment]
per @ZaradarBH suggest, we should check if Upgrade handler's version map is empty in mainnet:

- a full node
- modified binary version 1.0.4 with version map logging should do

**Comment by inon-man** [comment]
I will check empty version map on my node by logging it.

**Comment by inon-man** [comment]
I spinned up full node with pruned columbus-5 data and it has empty version map.

**Comment by edk208** [comment]
added block height guards, to trigger Feb 14th

**Review by ZaradarBH — APPROVED** [review]
